### ISSUE TITLE
Can you make it available for offline use?

### ISSUE THREAD

**Issue by mark522412** [apple1417/OpenHotfixLoader#13]
I don't always have internet so please if possible. thanks

**Comment by apple1417** [comment]
This is an eventual goal, but don't expect it any time soon, it's defeated several different people's attempts at reverse engineering (would be doing it if it were easy). As is we simply just need to get the existing hotfixes before we can inject our own.

**Comment by ricochhet** [comment]
@apple1417 I assume there isn't and/or won't be any updates on this?

I figured it would be possible to call the function(s) that handle the hotfixes directly, or fake a connection for long enough to load hotfixes. Easier said than done, I know. 😅

I became interested when I noticed this "quirk" with the hotfix system myself. I don't have much knowledge on the subject outside of networking, unfortunately. However, it seems like the game doesn't even make any attempts at connecting to anything while offline. I always enjoy projects related to game preservation, so it's a bit unfortunate for me to see that this is not easily solvable.

**Comment by apple1417** [comment]
Nobody's looked into it. We're all aware it's a bit of a problem, and occasionally people like yourself pop in with interest, but no one's had both the RE skills and the motivation to solve it. Fwiw gearbox still hosts the hotfixes for BL2, which is 14 years old at this point.

The hooks this tool uses to inject hotfixes are specifically the json parser functions - it injects deserialized json objects. That's the first "problem", we're not faking a full response, they don't even get called if it can't connect. Older hotfix injection methods were just a plain MitM, so would't have this problem. It's relatively easy to go from those jooks to the spot making the requests though, so with a bit of effort you could get past it.

The second problem, which also affected MitM-based injectors, is that the game won't even try request hotfixes until it finishes authenticating with SHiFT. Hotfixes aren't even behind a protected endpoint, you can view them [here](https://discovery.services.gearboxsoftware.com/v2/client/steam/pc/oak/verification), but it still won't even try. And no one's worked out how to fake auth.

Now there are other options, other than trying to fake the full networking flow. There are some hardcoded hotfixes in the exe - those somehow work without internet. Most hotfix mods are just setting a certain property to a fixed value, you could convert those to a Pak mod (some hotfixes rely on dynamically switching, they couldn't). Or if you write an SDK mod that parses hotfix syntax and applies them at the right time, just need to find the right hooks. None of these have been explored.

**Comment by sandgray** [comment]
@apple1417   Благодарю за расширенный ответ. Теперь понятно, почему все моды работают только он-лайн SHiFT.

Но всё же если SHiFT off-line - может быть в таком случае в клиенте можно отключить вообще проверку ключей - вобще всю процедуру авторизации в клиенте исключитть из процесса?

Я просто в шоке.  =(
Я вчера купил лицензию - именно для того, что бы играть с модами! И вчера несколько часов поиграл в своё удовольствие. (использовал ваш лоадер, так как на тот момент, в нём показалось проще сделать нужный мне порядок файлов).

А сегодня с утра игра играет, но SHiFT off-line! (при этом сайт грузиться и работает быстро. Кстати **_вопрос - игра пытается подключиться к SHifT по адресу shift.gearboxsoftware.com - или другому?_**). 

Я убил несколько часов на поиск решения проблемы SHiFT off-line только для  того, что бы узнать, что эта проблема уже наверное 5 лет и никто не имеет 100% способа её решения!

**Comment by sandgray** [comment]
Мне хочется матерится!

Я нашёл инструкцию, беглое прочтение которой внушало кое-какие надежды. https://www.reddit.com/r/borderlands3/comments/ig24e7/hotfix_modding_tutorial_easier_version/

И вот когда я скачал нужный софт я решил проверить в последний раз статус SHiFT -  SHiFT on-line и моды работают!  

Прошу прощения - просто эмоции - была куча свободного времени, которое планировал потратить на игру в своё удовольствие, но вместо этого убил на поиск инфы о том, что этой проблеме уже почти 5 лет  =(

**Comment by apple1417** [comment]
Yes SHiFT occasionally goes down for a few hours, disabling all hotfix mods, which is unfortunate. Gearbox used to announce these ahead of time. I'm sure if there was risk of it going down permanently, we'd see more people looking into this.

> But still, if SHiFT is off-line, maybe in this case it is possible to disable key verification in the client altogether - in general, exclude the entire authorization procedure in the client from the process?

This is the rough idea - someone just needs to reverse engineer it.
